File: playwright-agent/src/generator.ts
import type {
  DomMap,
  GeneratedTest,
  Requirement,
  TestGenerator,
  TestPlan,
} from "@qualityforge/core";

function locatorLine(dom: DomMap): string {
  const locator = dom.locators[0];
  if (locator?.role && locator.name) {
    return `    await expect(page.getByRole('${locator.role}', { name: ${JSON.stringify(locator.name)} })).toBeVisible();`;
  }
  if (locator?.label) {
    return `    await expect(page.getByLabel(${JSON.stringify(locator.label)})).toBeVisible();`;
  }
  return `    await expect(page.getByRole('main')).toBeVisible();`;
}

export class PlaywrightTestGenerator implements TestGenerator {
  async generate(
    requirement: Requirement,
    plan: TestPlan,
    dom: DomMap,
  ): Promise<GeneratedTest[]> {
    const url = dom.url ?? "about:blank";
    const spec = `import { test, expect } from '@playwright/test';

test(${JSON.stringify(requirement.title)}, async ({ page }) => {
    await page.goto(${JSON.stringify(url)});
${locatorLine(dom)}
});
`;

    return [
      {
        id: `${requirement.id}-e2e-1`,
        title: requirement.title,
        spec,
        locators: dom.locators,
      },
    ];
  }
}

File: backend/src/qualityforge/agents/generator.py
from qualityforge.models import DomMap, GeneratedTest, Requirement, TestPlan


def _locator_line(dom: DomMap) -> str:
    locator = dom.locators[0] if dom.locators else None
    if locator and locator.role and locator.name:
        return (
            f"    await expect(page.get_by_role({locator.role!r}, name={locator.name!r})"
            ").to_be_visible()"
        )
    if locator and locator.label:
        return f"    await expect(page.get_by_label({locator.label!r})).to_be_visible()"
    return "    await expect(page.get_by_role('main')).to_be_visible()"


def generate_tests(requirement: Requirement, plan: TestPlan, dom: DomMap) -> list[GeneratedTest]:
    del plan
    url = dom.url or "about:blank"
    spec = f'''from playwright.async_api import expect, Page

async def test_{requirement.id.replace("-", "_").lower()}(page: Page) -> None:
    await page.goto({url!r})
{_locator_line(dom)}
'''
    return [
        GeneratedTest(
            id=f"{requirement.id}-e2e-1",
            title=requirement.title,
            spec=spec,
            locators=dom.locators,
        )
    ]
